lfs/client.go

```diff
@@ -125,6 +126,58 @@ func (b *byteCloser) Close() error {
 	return nil
 }
 
+type upload struct {
+	OIDPath  string
+	Filename string
+	CB       CopyCallback
+}
+
+type UploadQueue struct {
+	uc     chan upload
+	ec     chan *WrappedError
+	wg     sync.WaitGroup
+	errors []*WrappedError
+}
+
+func NewUploadQueue(workers int) *UploadQueue {
+	q := &UploadQueue{uc: make(chan upload, workers), ec: make(chan *WrappedError)}
+
+	go func() {
+		for err := range q.ec {
+			q.errors = append(q.errors, err)
+		}
+	}()
+
+	for i := 0; i < workers; i++ {
+		go func(n int) {
+			for upload := range q.uc {
+				q.wg.Add(1)
+				err := Upload(upload.OIDPath, upload.Filename, upload.CB)
+				if err != nil {
+					q.ec <- err
+				}
+				q.wg.Done()
+			}
+		}(i)
+	}
+
+	return q
+}
+
+func (q *UploadQueue) Upload(oidPath, filename string, cb CopyCallback) {
+	q.uc <- upload{oidPath, filename, cb}
+}
+
+func (q *UploadQueue) Wait() {
+	close(q.uc)
+	q.wg.Wait()
+	close(q.ec)
```

Comment:
Correct on the goroutines.

For `Wait()`, starting or running is implicit, we don't need to wait until the loop that figures out what to upload to finish before starting uploads. See http://golang.org/pkg/os/exec/#Cmd.Wait and http://golang.org/pkg/sync/#WaitGroup.Wait (which is what it's actually doing).